### ISSUE TITLE
No touchpad (libinput) scroll on sway WM (wayland)

### ISSUE THREAD

**Issue by ghost** [vhakulinen/gnvim#70]
Is there a setting to enable scrolling with touchpad? I can select with the mouse, but not scroll. Thanks!

**Comment by badosu** [comment]
Do you have `set mouse=a` on your `init.vim`?

**Comment by lifranc** [comment]
I do. Tried without, no luck! Using sway if that's relevant.

**Comment by smolck** [comment]
I am using sway as well, and scrolling works with the mouse wheel, but not the touchpad. (That is not the case with the TUI when `set mouse=a` is on.)

**Comment by smolck** [comment]
Tested this on xorg/i3-gaps, and the problem does not occur; possibly a problem specific to wayland. @lifranc Would you be able test this [neovim-gtk](with https://github.com/daa84/neovim-gtk)?

**Comment by lifranc** [comment]
> Would you be able test this with [neovim-gtk](https://github.com/daa84/neovim-gtk)?

@smolck Sure will, but it'll have to wait for a few days now.

**Comment by smolck** [comment]
Tested it with neovim-gtk, and scrolling worked perfectly fine with the touchpad. Seems to be a problem with GNvim. I'll look into it.

**Comment by smolck** [comment]
It seems that GNvim isn't getting any input from the touchpad at all. I put `dbg!(e.get_direction())` [here]() but it didn't output anything when I used the touchpad to scroll (on sway), while it did output the scroll direction when I used the scroll wheel on an external mouse.

From that, this seems to be a GTK problem, but more debugging is probably necessary to be certain. (If it is a bug with GTK not getting the touchpad input on sway, best we could do would probably be to open an issue upstream to gtk-rs and then find a workaround. This isn't fixed in #71 for me, so it may be a bug that hasn't been noticed upstream yet.)

It'd probably be worthwhile to look into how neovim-gtk handles scroll events to see why it doesn't have this problem (it didn't when I tested it, at least).

**Comment by lifranc** [comment]
Has anyone found out more info on this? Terminal emulators are driving me crazy enough to give this some time.

**Comment by smolck** [comment]
@lifranc Other than the short debugging I did when you first posted this, I have not looked into it. (Most notably because pango 1.44 has made GNvim unusable for me; hence #124.) But If you would like to try your hand at some debugging or creating a PR, I'm happy to help by answering any questions you may have.

>  I put dbg!(e.get_direction()) here but it didn't output anything when I used the touchpad to scroll (on sway),

As a start if you're interested, I believe the link I forgot to put in the quoted portion above should have been in this function: https://github.com/vhakulinen/gnvim/blob/7b89a6de910252acf86c5035155b9df63fd46f31/src/ui/grid/grid.rs#L181-L202
That function may help you whilst debugging. It doesn't *seem* like GNvim isn getting any scrolling events from Sway/Wayland, but since this issue doesn't exist in Neovim-GTK I somewhat doubt that's the case. 

Since Neovim-GTK doesn't have this issue, maybe try looking through the files in this search: https://github.com/daa84/neovim-gtk/search?utf8=%E2%9C%93&q=Scroll&type= I'm not sure if any of that will help you find why this discrepancy between Neovim-GTK and GNvim is present (it's just a search of "Scroll" in Neovim-GTK), but I figured I'd still mention it.

Hopefully that helps ;) Not sure if you want to try this yourself, but if you do please go for it. Again, I'm happy to help where I can (I might look around a bit and see what I can find. Just not sure when I'll take the time to properly sit down and debug this).

**Comment by jeanguyomarch** [comment]
Hi, I'm having the same problem, but I think I've pinpoined its origin. I can indeed see events like *mouse drag*, but not *mouse scroll*. It seems this is caused by **smooth scrolling**, which is not captured by `gdk::EventBox`, as seen [here](https://github.com/vhakulinen/gnvim/blob/master/src/ui/grid/grid.rs#L111).
Changing this code to :

```rust
        eb.add_events(EventMask::SCROLL_MASK | EventMask::SMOOTH_SCROLL_MASK);
```

makes "something" to happen: the window scrolls down (very fast).
This is because of the handling of [ScrollDirection enumeration here](https://github.com/vhakulinen/gnvim/blob/master/src/ui/grid/grid.rs#L210-L213). We can change the code a bit to distinguish *up* and *down* when smooth scrolling is enabled (and ignore scrolling when neither up nor down is requested):

```rust
            let dir = match e.get_direction() {
                gdk::ScrollDirection::Up => ScrollDirection::Up,
                gdk::ScrollDirection::Down => ScrollDirection::Down,
                gdk::ScrollDirection::Smooth => {
                    let (_, dy) = e.get_scroll_deltas().unwrap();
                    if dy < 0.0 {
                        ScrollDirection::Up
                    } else if dy > 0.0 {
                        ScrollDirection::Down
                    } else {
                      return Inhibit(false);
                    }
                },
                _ => { return Inhibit(false); },
            };
```

Unfortunately, because scrolling is *smooth*, there are a **lot** of events that are passed directly to [nvim_input_mouse()](https://github.com/vhakulinen/gnvim/blob/master/src/ui/state.rs#L1010), which does not seem to provide some kind of "scroll speed" (as per [neovim's docs](https://neovim.io/doc/user/api.html)). This makes scrolling practically unusable.

I'm not sure on how to properly handle the speed of smooth scrolling. I haven't found how to disable it for a single application. If you have ideas, I can try them out. I'll let you know if I come up with a not-so-bad solution from my own.

**Comment by vhakulinen** [comment]
Fixed in #175.
